library/core/src/iter/traits/iterator.rs

```diff
@@ -591,6 +591,31 @@ pub trait Iterator {
         Intersperse::new(self, separator)
     }
 
+    /// Places an element generated by `separator` between all elements.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// #![feature(iter_intersperse)]
+    ///
+    /// let src = ["Hello", "to", "all", "people"].iter().copied();
+    /// let mut separator = [" ❤️ ", " 😀 "].iter().copied().cycle();
+    ///
+    /// let result = src.intersperse_with(|| separator.next().unwrap()).collect::<String>();
+    /// assert_eq!(result, "Hello ❤️ to 😀 all ❤️ people");
```

Comment:
Love the example ^^

Reply:
Is it ok that we use three other parts of `Iterator` (`cycle`, `copied` and `collect`) to demonstrate `intersperse_with`? Have a look at the second test in `test_intersperse_with`, which is less sexy, but may be more to the point.

Reply:
I think that's fine. Especially showing this usage of `.collect::<String>()` is useful, as that's probably going to be a common use case for `.intersperse_with()`.

You could add that `.intersperse_with(|| { n *= 2; n })` example before this one, if you think that's useful.